Chapter 5: Python Code

Smit-create opened the pull request.

Smit-create commented:
Results from firm exit:
![firm_exit_1](https://user-images.githubusercontent.com/55887635/200248823-d98f1140-55a5-42ec-a2af-3a7465fda0f5.png)
![firm_exit_2](https://user-images.githubusercontent.com/55887635/200248977-677ec314-f702-4c70-9763-1fd0e8bd0db6.png)

jstac commented on source_code_py/firm_exit.py:
```diff
@@ -0,0 +1,105 @@
+    z_vals, Q = mc.state_values, mc.P
+    return Model(n=n, z_vals=z_vals, Q=Q, β=β, s=s)
+
+
```
jit compile?

jstac commented on source_code_py/firm_exit.py:
```diff
@@ -0,0 +1,105 @@
+    I = np.identity(n)
+    return np.linalg.solve((I - β * Q), z_vals)
+
+
```
jit compile?

jstac commented on source_code_py/firm_exit.py:
```diff
@@ -0,0 +1,105 @@
+    h = z_vals + β * np.dot(Q, v)
+    return np.maximum(s, h)
+
+
```
jit compile?

jstac reviewed (approved):
Many thanks @Smit-create , great work.

Perhaps it's worth jit compiling some of the functions indicated below, just so the speed is comparable with the Julia code?  (I realize that these problems are small and the code executes quickly anyway.)

Smit-create commented:
I modified the code slightly ( fixed some minor bugs) to make it jit compile and got the following results:
![american_option_1](https://user-images.githubusercontent.com/55887635/200902898-aa286231-b6b3-49c1-8aa2-bf3573cfb803.png)
![american_option_2](https://user-images.githubusercontent.com/55887635/200902915-0b4a4751-daaf-4f44-99ce-9813fd1fb546.png)

Smit-create commented:
@jstac Does this look good now?

jstac commented on source_code_py/american_option.py:
```diff
@@ -7,12 +7,26 @@
+
+
+@njit
+def _eval(t, i_w, i_z, T, K):
```
@Smit-create , can you please call this function `exit_reward` and put in a docstring that says "Payoff to exercising the option at time t."

jstac commented on source_code_py/american_option.py:
```diff
@@ -7,12 +7,26 @@
+
+
+@njit
+def product(a, b):
```
Can this function be replaced by something in https://quanteconpy.readthedocs.io/en/latest/tools/gridtools.html, such as `cartesian`?

Smit-create replied:
Yes, that's the same, but using that will throw jit errors as:
```
Untyped global name 'cartesian': Cannot determine Numba type of <class 'function'>

File "american_option.py", line 56:
def C(h, model):
    <source elided>
    z_idx, w_idx = np.arange(len(z_vals)), np.arange(len(w_vals))
    for (t, i_z) in cartesian(t_vals, z_idx):
```

jstac commented:
Thanks @Smit-create , please see my comments above.

jstac commented:
Thanks @Smit-create .  I wonder if we should modify `cartesian` to be jit-aware --- otherwise we will need to repeat your `product` function in quite a few files for this project.  It's just a suggestion and I'm open to other ideas.

Regarding the figures, I notice that the font size for "in the money" and "out of the money" is too large in the Python code (please see above).

Finally, for successive approximation, perhaps we should use https://quanteconpy.readthedocs.io/en/latest/tools/compute_fp.html

Smit-create commented:
> Finally, for successive approximation, perhaps we should use https://quanteconpy.readthedocs.io/en/latest/tools/compute_fp.html

I was trying to avoid using direct library functions so that if anyone tries to compare python and Julia code, it would be easy to relate the things easily.

Smit-create commented:
Fixed the plot
![american_option_2](https://user-images.githubusercontent.com/55887635/201506950-cda85c50-120f-415d-86b2-399b87e82734.png)

Smit-create commented:
> I wonder if we should modify `cartesian` to be jit-aware

We can try to make that change in the `quantecon` itself? I mean we can try to make the functions/classes jit-aware as much as we can?

Smit-create commented:
Thanks, @jstac. I have addressed your comments. If we replace `successive_approx` with `compute_fp` then we will need to do that in all the previous code to maintain uniformity and that will make it differ slightly from Julia code. Should I change that? Or keep the current version?

jstac commented:
Thanks for the fixes @Smit-create 

> If we replace successive_approx with compute_fp then we will need to do that in all the previous code to maintain uniformity and that will make it differ slightly from Julia code. Should I change that? Or keep the current version?

I hadn't thought about this carefully enough.  How about this policy: we don't worry about matching Python and Julia code, as long as we reproduce the figures.

As a general rule, let's write the Python code as efficiently as possible, and with minimal repetition.  

Following this rule, let's use `compute_fp` here.  But there's no need to go back and change the earlier ones --- it might be good for people who are starting out to see all the details.

Smit-create commented:
> As a general rule, let's write the Python code as efficiently as possible, with minimal repetition.

Thanks, @jstac. I just replaced the `product` and instead used loops with `prange` and also used `compute_fp` from `quantecon`. This works very fast now.

jstac commented:
Very nice work @Smit-create .  I'll merge this now.
